lib/tds/connection.ex

```diff
@@ -270,45 +268,61 @@ defmodule Tds.Connection do
   defp command(:attn, s) do
     timeout = s.opts[:timeout] || @timeout
     attn_timer_ref = :erlang.start_timer(timeout, self(), :command)
-    Protocol.send_attn(%{s |attn_timer: attn_timer_ref, pak_header: "", pak_data: "", tail: "", state: :attn})
+    Protocol.send_attn(%{s | attn_timer: attn_timer_ref, pak_header: "", pak_data: "", tail: "", state: :attn})
   end
 
+  #no data to process
   defp new_data(<<_data::0>>, s) do
     {:ok, s}
   end
+
+  #DONE_ATTN The DONE message is a server acknowledgement of a client ATTENTION message.
   defp new_data(<<0xFD, 0x20, _cur_cmd::binary(2), 0::size(8)-unit(8), _tail::binary>>, %{state: :attn} = s) do
     s = %{s | pak_header: "", pak_data: "", tail: ""}
     Protocol.message(:attn, :attn, s)
   end
+
+  #shift 8 bytes while in attention state, Protocol updates state
   defp new_data(<<_b::size(1)-unit(8), tail::binary>>, %{state: :attn} = s), do: new_data(tail, s)
 
-  defp new_data(<<packet::binary>>, %{state: state, pak_data: buf_data, pak_header: buf_header, tail: tail} = s) do
-    <<type::int8, status::int8, size::int16, head_rem::int32, data::binary>> = tail <> packet
-    if buf_header == "" do
-      buf_header = <<type::int8, status::int8, size::int16, head_rem::int32>>
+  #no packet header yet
+  defp new_data(<<data::binary>>, %{pak_header: ""} = s) do
+    if byte_size(data) >= 8 do
+      #assume incoming data starts with packet header, if it's long enough
+      <<pak_header::binary(8), tail::binary>> = data
+      new_data(tail, %{s | pak_header: pak_header})
     else
-      data = tail <> packet
+      #have no packet header yet, wait for more data
+      {:ok, %{s | tail: data}}
     end
+  end
 
-    <<type::int8, status::int8, size::int16, _spid::int16, _pack_id::int8, _window::int8>> = buf_header
-    size = size - 8
+  defp new_data(<<data::binary>>, %{state: state, pak_header: pak_header, pak_data: pak_data} = s) do
+    <<type::int8, status::int8, size::int16, _head_rem::int32>> = pak_header
+    size = size - 8 #size includes packet header
 
     case data do
-      <<data :: binary(size), tail :: binary>> ->
+      <<data::binary(size), tail::binary>> ->
+        #satisfied size specified in packet header
         case status do
           1 ->
-            msg = Messages.parse(state, type, buf_header, buf_data<>data)
+            #status 1 means last packet of message
+            #TODO Messages.parse does not use pak_header
+            msg = Messages.parse(state, type, pak_header, pak_data <> data)
             case Protocol.message(state, msg, s) do
               {:ok, s} ->
-                new_data(tail, %{s | pak_header: "", pak_data: "", tail: tail})
```

Comment:
notice that this was setting tail and passing tail, which would effective duplicate tail in next run (tail <> packet). not sure how it didn't fail more often? most queries only have 1 message?

from what i can tell, tail only needs to be set when waiting for more data, tail is cleared before the only entry into new_data so it's always blank in new_data